lib/router/handler-info.js

```diff
@@ -36,22 +51,34 @@ HandlerInfo.prototype = {
     return this.params || {};
   },
 
+  updateHandler: function(handler) {
+    // Store the name of the handler on the handler for easy checks later
+    handler._handlerName = this.name;
+    return this.handler = handler;
+  },
+
   resolve: function(shouldContinue, payload) {
     var checkForAbort  = bind(this, this.checkForAbort,      shouldContinue),
         beforeModel    = bind(this, this.runBeforeModelHook, payload),
         model          = bind(this, this.getModel,           payload),
         afterModel     = bind(this, this.runAfterModelHook,  payload),
-        becomeResolved = bind(this, this.becomeResolved,     payload);
-
-    return Promise.resolve(undefined, this.promiseLabel("Start handler"))
-           .then(checkForAbort, null, this.promiseLabel("Check for abort"))
-           .then(beforeModel, null, this.promiseLabel("Before model"))
-           .then(checkForAbort, null, this.promiseLabel("Check if aborted during 'beforeModel' hook"))
-           .then(model, null, this.promiseLabel("Model"))
-           .then(checkForAbort, null, this.promiseLabel("Check if aborted in 'model' hook"))
-           .then(afterModel, null, this.promiseLabel("After model"))
-           .then(checkForAbort, null, this.promiseLabel("Check if aborted in 'afterModel' hook"))
-           .then(becomeResolved, null, this.promiseLabel("Become resolved"));
+        becomeResolved = bind(this, this.becomeResolved,     payload),
+        self = this;
+
+    return Promise.resolve(this.handlerPromise, this.promiseLabel("Start handler"))
+            .then(function(handler) {
+              return Promise.resolve(handler)
```

Comment:
So we can do that, the one issue with that approach (and the reason the code is this way) is if we have an error early on in the chain (for example [when checking `inaccessibleByURL`](https://github.com/tildeio/router.js/pull/183/files#diff-5863acacff50c33c3f6077ebdf56d4fbR53)), then we have to propagate it through the whole chain which is annoying.